server/src/main/java/org/elasticsearch/index/mapper/FieldDataParseHelper.java

```diff
@@ -0,0 +1,357 @@
+                try (XContentBuilder builder = XContentBuilder.builder(parser.contentType().xContent())) {
+                    builder.copyCurrentStructure(parser);
+                    return encode(builder);
+                }
+            }
+
+            @Override
+            void decodeAndWrite(XContentBuilder b, BytesRef r) throws IOException {
+                switch (r.bytes[r.offset]) {
+                    case 'c' -> decodeAndWriteXContent(b, XContentType.CBOR, r);
+                    case 'j' -> decodeAndWriteXContent(b, XContentType.JSON, r);
+                    case 's' -> decodeAndWriteXContent(b, XContentType.SMILE, r);
+                    case 'y' -> decodeAndWriteXContent(b, XContentType.YAML, r);
+                    default -> throw new IllegalArgumentException("Can't decode " + r);
+                }
+            }
+        };
+
+        TypeUtils(char encoding) {
+            this.encoding = encoding;
+        }
+
+        byte getEncoding() {
+            return (byte) encoding;
+        }
+
+        final char encoding;
+
+        abstract StoredField buildStoredField(String name, XContentParser parser) throws IOException;
+
+        abstract byte[] encode(XContentParser parser) throws IOException;
+
+        abstract void decodeAndWrite(XContentBuilder b, BytesRef r) throws IOException;
+
+        static byte[] encode(BigInteger n, Byte encoding) throws IOException {
+            byte[] twosCompliment = n.toByteArray();
+            byte[] encoded = new byte[1 + twosCompliment.length];
+            encoded[0] = encoding;
+            System.arraycopy(twosCompliment, 0, encoded, 1, twosCompliment.length);
+            return encoded;
+        }
+
+        static byte[] encode(BigDecimal n, Byte encoding) {
+            byte[] twosCompliment = n.unscaledValue().toByteArray();
+            byte[] encoded = new byte[5 + twosCompliment.length];
+            encoded[0] = 'd';
+            ByteUtils.writeIntLE(n.scale(), encoded, 1);
+            System.arraycopy(twosCompliment, 0, encoded, 5, twosCompliment.length);
+            return encoded;
+        }
+
+        static byte[] encode(byte[] b) {
+            byte[] encoded = new byte[1 + b.length];
+            encoded[0] = 'b';
+            System.arraycopy(b, 0, encoded, 1, b.length);
+            return encoded;
+        }
+
+        static byte[] encode(XContentBuilder builder) throws IOException {
+            BytesReference b = BytesReference.bytes(builder);
+            byte[] encoded = new byte[1 + b.length()];
+            encoded[0] = switch (builder.contentType()) {
+                case JSON -> 'j';
+                case SMILE -> 's';
+                case YAML -> 'y';
+                case CBOR -> 'c';
+                default -> throw new IllegalArgumentException("unsupported type " + builder.contentType());
+            };
+
+            int position = 1;
+            BytesRefIterator itr = b.iterator();
```

Comment:
This was copy-pasted from IgnoreMalformedStoredValues. I think the idea is to avoid the double copy-paste, as `toBytes` may copy as well. It's unlikely to have any significant impact, but it's probably better to make no unnecessary changes here.

Reply:
Ah, makes sense.